### PR TITLE
new fidelity grad and operator fidelity grad.

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Not working until https://github.com/QuantumBFS/LuxurySparse.jl/pull/29 is merged.

File: src/measure_ops.jl [hunk 1]
```diff
@@ -19,6 +19,50 @@ function eigenbasis(op::AbstractBlock{N}) where {N}
...
+                return false
+            else
+                occ[i] = true
+            end
+        end
+    end
+    return true
+end
+
+function eigenbasis(op::ChainBlock{N}) where N
```
Comment (GiggleLiu) [inline src/measure_ops.jl:46]: @wangleiphy Is this ok?

Comment (wangleiphy) [reply src/measure_ops.jl:46]: It is OK! I tested it in my application. Works and is faster. 
👍

**Review by wangleiphy — APPROVED** [review]

File: src/measure_ops.jl [hunk 2]
```diff
@@ -19,6 +19,50 @@ function eigenbasis(op::AbstractBlock{N}) where {N}
...
+    @eval function eigenbasis(op::$BT)
+        @warn "eigenbasis on blocktype `$($BT)` calls into the fallback implementation, which might be slow. Try using `kron`, `repeat` if items commute to each oher."
+        invoke(eigenbasis, Tuple{AbstractBlock}, op)
+    end
+end
+
+"""
+Return true if operators commute to each other.
+"""
+function simple_commute_eachother(ops::Vector{<:AbstractBlock{N}}) where N
```
Comment (Roger-luo) [inline src/measure_ops.jl:32]: what's the difference between this an is commute?

File: src/composite/control.jl [hunk 3]
```diff
@@ -210,8 +210,10 @@ function Base.copy(ctrl::ControlBlock{N,BT,C,M}) where {BT,N,C,M}
 end
 
 function YaoBase.iscommute(x::ControlBlock{N}, y::ControlBlock{N}) where {N}
-    if x.locs == y.locs && x.ctrl_locs == y.ctrl_locs
```
Comment (Roger-luo) [inline src/composite/control.jl:213]: why this is removed?

**Review by Roger-luo — COMMENTED** [review]
and can you move `hadmard_product` to `utils.jl` for now? I don't think it is necessary to stay in LuxurySparse, I can support broadcast once I have time

File: src/composite/control.jl (hunk 3 above)
Comment (GiggleLiu) [reply src/composite/control.jl:213]: to commute two control blocks, ctrl locs are not nessesarily same.

File: src/measure_ops.jl (hunk 2 above)
Comment (GiggleLiu) [reply src/measure_ops.jl:32]: this line checks address conflicts, which is from practical considerations. is_commute should have stronger power in simplification of eigenbasis, however it is  harder to analyse.

**Pull request merged**